molt/src/commands.rs

```diff
@@ -1121,6 +1121,31 @@ pub fn cmd_string_equal(_interp: &mut Interp, _: ContextID, argv: &[Value]) -> M
     }
 }
 
+/// string first *needleString* *haystackString* ?*startIndex*?
+pub fn cmd_string_first(_interp: &mut Interp, _: ContextID, argv: &[Value]) -> MoltResult {
+    check_args(2, argv, 4, 5, "needleString haystackString ?startIndex?")?;
+
+    let needle = argv[2].as_str();
+    let haystack = argv[3].as_str();
+
+    let start = if argv.len() == 5 {
+        argv[4].as_int()?
+    } else {
+        0
+    };
+
+    let pos = if start < 0 || start as usize >= haystack.len() {
+        -1
+    } else {
+        haystack[start as usize..]
```

Comment:
I think there's a problem here.  If I'm not mistaken, you are starting the search at the nth byte, not the nth character.  There will be problems if `start` addresses the second or third byte of a multi-byte character; at the very least it won't get the right answer.  Unless you know something I don't?  Though of course it will work fine if all characters are one byte.

I think what you probably need to do is take `haystack.chars()`, skip the first `start` characters to create a new iterator, and then convert the head of the iterator back to a `&str` and do the find in that.

Reply:
You are right.  Not only did I use the byte length and byte indices for `startIndex` and `lastIndex`, I also *returned* a byte index from `string first` and `string last`.  That is what `find()` and `rfind()` return.  I have rewritten `string first`/`last` to operate on chars and to return char indices.  I have also added tests that panic if the commands mishandle multibyte characters and fail if they return byte indices.  The Rust idiom for mapping a character index to a byte index seems to be `s.char_indices().nth(n)`.